source/common/common/bit_array.h

```diff
@@ -0,0 +1,147 @@
+#pragma once
+
+#include <cstdint>
+#include <cstring>
+
+#include "source/common/common/assert.h"
+#include "source/common/common/safe_memcpy.h"
+
+namespace Envoy {
+
+/**
+ * BitArray is an fixed sized array of fixed width bits. As such once
+ * constructed it will not grow in the number of elements it can hold, or change
+ * the number of bits an element is considered.
+ *
+ * At a minimum this will allocate 2 * word_size worth of bytes. To simplify the
+ * implementation for accessing data in the last word, we allocate an additional
+ * word_size of bytes beyond what is necessary to hold the items in the BitArray.
+ *
+ * Methods are deliberately implemented in the header to hint to the compiler to
+ * inline.
+ *
+ * Limitations to this class are:
+ *  1) It is thread compatible.
+ *  2) It assumes a 64-bit architecture.
+ *  3) It works with a maximum width of 32-bits
+ */
```

Comment:
This is probably a dumb question but isn't this https://en.cppreference.com/w/cpp/utility/bitset? If it's not can you add comments on how it's different?

Reply:
From comment above: " BitArray is an fixed sized array of fixed width bits."

Note the "fixed width bits", entries can range from being 1-32 bits each. For bit set elements are just 1 bit width. From looking at get: https://en.cppreference.com/w/cpp/utility/bitset/operator_at we can see it just rets a bool

Reply:
OK, got it. Pedantically I found this a bit confusing. Maybe something like "BitArray is a fixed sized array of fixed with bit sets" or something like that, or just a note to clarify how it's different from bitset. As an aside, I assume that in some cases `std::array<std::bitset>` or similar would not compress as well, but would that end up a simpler implementation? Anyway I would just have some note about bitset to make it more clear for quick readers.

Reply:
sg, added a note to clarify how it is different from `std::bitset`.

Yea the compression is __significantly worse__ in all cases except when using bit array for 32-bit elements.
```
std::bitset<1> foo;
std::cout << "Size of foo: " << sizeof(foo) << std::endl;
```

Yields 4, which would be 32x  more than BitArray representation. Moreover we'd need to instantiate templates for `bitset<1>, bitset<2>, ..., bitset<32>`. which seems like an easy way to bloat the binary.